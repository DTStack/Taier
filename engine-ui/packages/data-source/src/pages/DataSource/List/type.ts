/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

//分页字段
export interface IPagination {
  currentPage: number;
  pageSize: number;
}
export interface IOther {
  search: string;
  dataTypeList: string[];
  appTypeList: number[];
  isMeta: number;
  status: number[];
}

//列表操作字段
export interface IRecord {
  dataInfoId: number;
  isAuth: number;
}
