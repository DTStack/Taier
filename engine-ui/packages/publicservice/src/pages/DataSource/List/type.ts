//分页字段
export interface IPagination {
  currentPage: number;
  pageSize: number;
}
export interface IOther {
  search: string;
  dataTypeList: string[];
  appTypeList: number[];
  isMeta: number;
  status: number[];
}

//列表操作字段
export interface IRecord {
  dataInfoId: number;
  isAuth: number;
}
