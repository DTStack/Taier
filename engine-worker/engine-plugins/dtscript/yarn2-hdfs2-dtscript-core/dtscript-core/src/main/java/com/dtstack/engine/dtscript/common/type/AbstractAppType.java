package com.dtstack.engine.dtscript.common.type;


import com.dtstack.engine.dtscript.DtYarnConfiguration;
import com.dtstack.engine.dtscript.client.ClientArguments;
import org.apache.commons.lang.StringUtils;
import org.apache.hadoop.yarn.conf.YarnConfiguration;
import org.codehaus.jackson.map.ObjectMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;
import java.util.Map;

public abstract class AbstractAppType {

    protected static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();

    protected static final Logger buildCmdLog = LoggerFactory.getLogger(AbstractAppType.class);

    public static AbstractAppType fromString(String type) {
        if (StringUtils.isBlank(type)) {
            return new DummyType();
        } else if (type.equalsIgnoreCase(AppTypeEnum.SHELL.name())) {
            return new ShellType();
        } else if (type.equalsIgnoreCase(AppTypeEnum.PYTHON.name()) || type.equalsIgnoreCase(AppTypeEnum.PYTHON2.name())) {
            return new Python2Type();
        } else if (type.equalsIgnoreCase(AppTypeEnum.PYTHON3.name())) {
            return new Python3Type();
        } else if (type.equalsIgnoreCase(AppTypeEnum.JLOGSTASH.name())) {
            return new JLogstashType();
        } else if (type.equalsIgnoreCase(AppTypeEnum.JUPYTER.name())) {
            return new JupyterType();
        }
        throw new IllegalArgumentException("Unsupported appType: " + type);
    }

    public String cmdPrefix(YarnConfiguration config) {
        return "";
    }

    public String buildCmd(ClientArguments clientArguments, YarnConfiguration conf) {
        if (StringUtils.isNotBlank(clientArguments.getLaunchCmd())) {
            return clientArguments.getLaunchCmd();
        } else {
            String fullPath = clientArguments.getFiles()[0];
            String[] parts = fullPath.split("/");
            String encodedOpts = "";

            return cmdPrefix(conf) + " " + parts[parts.length - 1] + " " + encodedOpts;
        }
    }

    /**
     * 每个类型，可能对执行时的命令有额外的参数处理
     */
    public String cmdContainerExtra(String cmd, DtYarnConfiguration conf, Map<String, Object> containerInfo) {
        return cmd;
    }

    abstract public String name();

    public void env(List<String> envList) {
        envList.add("LANG=zh_CN.UTF-8");
        envList.add("CLASSPATH=" + "./:" + System.getenv("CLASSPATH"));
        envList.add("PATH=" + "./:" + System.getenv("PATH"));
    }

}