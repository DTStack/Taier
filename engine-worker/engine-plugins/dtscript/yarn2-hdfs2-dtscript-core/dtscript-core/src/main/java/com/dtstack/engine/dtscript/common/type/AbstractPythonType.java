package com.dtstack.engine.dtscript.common.type;

import com.dtstack.engine.dtscript.client.ClientArguments;
import org.apache.commons.lang.StringUtils;
import org.apache.hadoop.yarn.conf.YarnConfiguration;

import java.net.URLEncoder;
import java.util.List;

public abstract class AbstractPythonType extends AbstractAppType {

    @Override
    public void env(List<String> envList) {
        super.env(envList);
    }

    @Override
    public String buildCmd(ClientArguments clientArguments, YarnConfiguration conf) {
        if (StringUtils.isNotBlank(clientArguments.getLaunchCmd())) {
            return StringUtils.replaceOnce(clientArguments.getLaunchCmd(), "python", cmdPrefix(conf));
        }

        String cmdOpts = "";
        if (StringUtils.isNotBlank(clientArguments.getCmdOpts())) {
            try {

                cmdOpts = clientArguments.getCmdOpts();
                if (clientArguments.getLocalFile()){
                    cmdOpts += " " + clientArguments.getApplicationId();
                }
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        }

        String execFile =clientArguments.getFiles()[0];
        if (!clientArguments.getLocalFile()){
            String fullPath = clientArguments.getFiles()[0];
            String[] parts = fullPath.split("/");
            execFile = parts[parts.length - 1];
        }
        return cmdPrefix(conf) + " " + execFile + " " + cmdOpts;
    }
}
