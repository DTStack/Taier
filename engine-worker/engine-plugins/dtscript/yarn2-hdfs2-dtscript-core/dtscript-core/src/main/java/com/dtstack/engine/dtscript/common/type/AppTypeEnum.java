package com.dtstack.engine.dtscript.common.type;

/**
 * company: www.dtstack.com
 * author: toutian
 * create: 2019/7/30
 */
public enum AppTypeEnum {
    //JLOGSTASH
    JLOGSTASH,
    //SHELL
    SHELL,
    //PYTHON
    PYTHON,
    //PYTHON2
    PYTHON2,
    //PYTHON3
    PYTHON3,
    //JUPYTER
    JUPYTER
}
