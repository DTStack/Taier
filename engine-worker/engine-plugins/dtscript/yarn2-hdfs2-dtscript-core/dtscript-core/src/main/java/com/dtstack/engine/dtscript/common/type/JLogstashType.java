package com.dtstack.engine.dtscript.common.type;

import ch.qos.logback.classic.Level;
import com.dtstack.engine.common.exception.ExceptionUtil;
import com.dtstack.engine.dtscript.DtYarnConfiguration;
import com.dtstack.engine.dtscript.client.ClientArguments;
import com.dtstack.engine.dtscript.util.Base64Util;
import com.dtstack.engine.dtscript.util.GZipUtil;
import com.dtstack.engine.dtscript.util.NetUtils;
import org.apache.commons.collections.MapUtils;
import org.apache.commons.lang.StringUtils;
import org.apache.hadoop.yarn.conf.YarnConfiguration;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;


public class JLogstashType extends AbstractAppType {

    private static final Logger LOG = LoggerFactory.getLogger(JLogstashType.class);

    @Override
    public String buildCmd(ClientArguments clientArguments, YarnConfiguration conf) {

        String root = conf.get("jlogstash.root");
        if (StringUtils.isBlank(root)) {
            throw new IllegalArgumentException("Must specify jlogstash.root");
        }
        String javaHome = conf.get("java.home", "");
        if (StringUtils.isNotBlank(javaHome) && !javaHome.endsWith("/")) {
            javaHome += "/";
        }
        String cmdOpts = clientArguments.getCmdOpts();
        if (StringUtils.isBlank(cmdOpts)) {
            throw new IllegalArgumentException("Must specify cmdOpts");
        }
        String appName = clientArguments.getAppName();
        if (StringUtils.isBlank(appName)) {
            appName = "jlogstashJob";
        }

        if (buildCmdLog.isDebugEnabled()) {
            buildCmdLog.debug("encodedOpts: " + clientArguments.getCmdOpts());
            buildCmdLog.debug("Building jlogstash launch command");
        }

        List<String> jlogstashArgs = new ArrayList<>(20);
        jlogstashArgs.add(javaHome + "java");
        jlogstashArgs.add(clientArguments.getJvmOpts());
        jlogstashArgs.add("-Xms" + (clientArguments.getWorkerMemory() - clientArguments.getWorkerReservedMemory()) + "m");
        jlogstashArgs.add("-Xmx" + (clientArguments.getWorkerMemory() - clientArguments.getWorkerReservedMemory())  + "m");
        jlogstashArgs.add("-cp " + root + "/jlogstash*.jar");
        jlogstashArgs.add("com.dtstack.jlogstash.JlogstashMain");
        jlogstashArgs.add("-l stdout");
        jlogstashArgs.add("-" + getJlogstashLogLevel(clientArguments.getLogLevel().toUpperCase()));
        jlogstashArgs.add("-f " + clientArguments.getCmdOpts());
        jlogstashArgs.add("-p " + root);
        jlogstashArgs.add("-name " + appName);

        StringBuilder command = new StringBuilder();
        for (String arg : jlogstashArgs) {
            command.append(arg).append(" ");
        }
        if (buildCmdLog.isDebugEnabled()) {
            buildCmdLog.debug("jlogstash launch command: " + command.toString());
        }

        return command.toString();

    }

    private String getJlogstashLogLevel(String logLevel) {
        if (Level.TRACE.levelStr.equals(logLevel)) {
            return "vvvvv";
        } else if (Level.DEBUG.levelStr.equals(logLevel)) {
            return "vvvv";
        } else if (Level.INFO.levelStr.equals(logLevel)) {
            return "vvv";
        } else if (Level.WARN.levelStr.equals(logLevel)) {
            return "vv";
        } else if (Level.ERROR.levelStr.equals(logLevel)) {
            return "v";
        } else {
            return "vvv";
        }
    }

    /**
     * logstash 需要对inputs中存在beats的情况，为port进行模式处理
     * 如果端口存在，则使用端口；
     * 不存在，则使用默认端口6767。
     * <p>
     * 端口被占用会进行检测，并递增进行重新设置端口，如6767被占用，则使用6768
     */
    @Override
    public String cmdContainerExtra(String cmd, DtYarnConfiguration conf, Map<String, Object> containerInfo) {
        try {
            String[] args = cmd.split("\\s+");
            String fStr = null;
            int idx = -1;
            for (int i = 0; i < args.length - 1; ++i) {
                if ("-f".equals(args[i])) {
                    fStr = Base64Util.baseDecode(GZipUtil.deCompress(args[i + 1]));
                    idx = i + 1;
                    break;
                }
            }
            boolean isChg = false;
            if (StringUtils.isNotBlank(fStr)) {
                Map configs = OBJECT_MAPPER.readValue(fStr, Map.class);
                List<Map> inputs = (List<Map>) MapUtils.getObject(configs, "inputs", Collections.EMPTY_LIST);
                if (!inputs.isEmpty()) {
                    int i = 1;
                    for (Map input : inputs) {
                        Iterator<Map.Entry<String, Map>> inputIt = input.entrySet().iterator();
                        while (inputIt.hasNext()) {
                            Map.Entry<String, Map> inputEntry = inputIt.next();
                            String inputType = inputEntry.getKey();
                            Map<String, Object> inputConfig = inputEntry.getValue();
                            if ("Beats".equalsIgnoreCase(inputType)) {
                                isChg = true;
                                int configPort = MapUtils.getInteger(inputConfig, "port", 6767);
                                int port = NetUtils.getAvailablePortRange(configPort, 65535);
                                inputConfig.put("port", port);

                                Map<String, Object> beats = new HashMap<>(1);
                                beats.put("port", port);
                                containerInfo.put("beats" + i++, beats);
                            }
                        }
                    }
                }
                fStr = OBJECT_MAPPER.writeValueAsString(configs);
            }
            if (idx != -1 && isChg) {
                args[idx] = GZipUtil.compress(Base64Util.baseEncode(fStr));

                cmd = StringUtils.join(args, " ");
            }
        } catch (Exception e) {
            LOG.error("JLogstashType.cmdContainerExtra error:", e);
        }
        return super.cmdContainerExtra(cmd, conf, containerInfo);
    }

    @Override
    public String name() {
        return "JLOGSTASH";
    }

    @Override
    public void env(List<String> envList) {
        super.env(envList);
    }
}