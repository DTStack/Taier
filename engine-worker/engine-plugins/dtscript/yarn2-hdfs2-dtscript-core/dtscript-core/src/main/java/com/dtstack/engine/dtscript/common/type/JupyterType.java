/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.dtstack.engine.dtscript.common.type;


import com.dtstack.engine.dtscript.DtYarnConfiguration;
import com.dtstack.engine.dtscript.client.ClientArguments;
import com.dtstack.engine.dtscript.util.NetUtils;
import org.apache.commons.lang3.StringUtils;
import org.apache.hadoop.yarn.conf.YarnConfiguration;

import java.io.File;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

public class JupyterType extends AbstractAppType {

    private final static String JUPYTER_NOTEBOOK_CONFIG_PREFIX = "c.";
    private final static String JUPYTER_NOTEBOOK_CONFIG_SIGN = "=";

    private final static String JUPYTER_NOTEBOOK_CONFIG_PORT_TMP = "#{port}";

    private final static String LINEFEED = "\n";
    private final static String JUPYTER_PROJECT_ROOT = "jupyter.project.root";
    private final static String JUPYTER_PROJECT_DIR_KEY = "jupyter.project.dir";
    private final static String JUPYTER_CONFIG_DIR_KEY = "jupyter.conf.dir";
    private final static String JUPYTER_CONFIG_DIR_SUFFIX = "/conf";
    /** log file directory */
    private final static String JUPYTER_LOG_DIR_SUFFIX = "/log";
    private final static String JUPYTER_CONFIG_FILENAME = "jupyter_notebook_config.py";
    private final static String JUPYTER_BIN_PATH = "jupyter.path";
    private final static String JUPYTER_WORKSPACE_DIR_KEY = "jupyter.workspace.dir";
    private final static String JUPYTER_WORKSPACE_DIR_SUFFIX = "/workspace";
    private final static String JUPYTER_ALLOW_ROOT = "jupyter.allow.root";
    /** debug configuration*/
    private final static String JUPYTER_LOG_DEBUG = "jupyter.log.debug";
    /** log file name */
    private final static String JUPYTER_LOG_FILENAME = "jupyter.log";
    private final static String JUPYTER_BASE_URL_PREFIX = "/aiworks/jupyter/";
    private final static String[] DEFAULT_PORT_RANGE = new String[]{"8888", "65535"};


    @Override
    public String buildCmd(ClientArguments clientArguments, YarnConfiguration conf) {

        String jupyterBinPath = conf.get(JUPYTER_BIN_PATH, "");
        if (StringUtils.isBlank(jupyterBinPath)) {
            throw new IllegalArgumentException(JUPYTER_BIN_PATH + " must be set");
        }
        boolean allowRoot = conf.getBoolean(JUPYTER_ALLOW_ROOT, true);
        boolean allowDebug = conf.getBoolean(JUPYTER_LOG_DEBUG, false);

        String jupyterProjectRoot = conf.get(JUPYTER_PROJECT_ROOT, "");
        if (StringUtils.isBlank(jupyterProjectRoot)) {
            throw new IllegalArgumentException(JUPYTER_PROJECT_ROOT + " must be set");
        }
        if (!jupyterProjectRoot.endsWith("/")) {
            jupyterProjectRoot += "/";
        }
        String jupyterProject = jupyterProjectRoot + clientArguments.getAppName();
        String jupyterWorkspace = jupyterProject + JUPYTER_WORKSPACE_DIR_SUFFIX;
        String jupyterConfDir = jupyterProject + JUPYTER_CONFIG_DIR_SUFFIX;
        String jupyterLogDir = jupyterProject + JUPYTER_LOG_DIR_SUFFIX;
        String jupyterBaseUrl = JUPYTER_BASE_URL_PREFIX + clientArguments.getAppName();
        conf.set(JUPYTER_PROJECT_DIR_KEY, jupyterProject);
        conf.set(JUPYTER_WORKSPACE_DIR_KEY, jupyterWorkspace);
        conf.set(JUPYTER_CONFIG_DIR_KEY, jupyterConfDir);

        StringBuilder bashScript = new StringBuilder(400);
        bashScript.append(cmdPrefix(conf)).append(" <<JUPYTER").append(LINEFEED);
        bashScript.append("if [ ! -d \"").append(jupyterWorkspace).append("\" ]; then").append(LINEFEED);
        bashScript.append("  mkdir -p \"").append(jupyterWorkspace).append("\"").append(LINEFEED);
        bashScript.append("fi").append(LINEFEED);
        bashScript.append("if [ ! -d \"").append(jupyterConfDir).append("\" ]; then").append(LINEFEED);
        bashScript.append("  mkdir -p \"").append(jupyterConfDir).append("\"").append(LINEFEED);
        bashScript.append("fi").append(LINEFEED);
        bashScript.append("if [ ! -d \"").append(jupyterLogDir).append("\" ]; then").append(LINEFEED);
        bashScript.append("  mkdir -p \"").append(jupyterLogDir).append("\"").append(LINEFEED);
        bashScript.append("fi").append(LINEFEED);
        bashScript.append("cd \"").append(jupyterConfDir).append("\"").append(LINEFEED);
        bashScript.append(generateJupyterNotebookConfig(conf, jupyterWorkspace, jupyterBaseUrl));
        bashScript.append(jupyterBinPath).append(" --config ").append(jupyterConfDir).append("/").append(JUPYTER_CONFIG_FILENAME);
        if (allowRoot) {
            bashScript.append(" --allow-root");
        }
        if (allowDebug) {
            bashScript.append(" --debug");
        }
        bashScript.append(" > ").append(jupyterLogDir).append("/").append(JUPYTER_LOG_FILENAME).append(" 2>&1 ");
        bashScript.append(LINEFEED);
        bashScript.append("JUPYTER");
        return bashScript.toString();
    }

    @Override
    public String cmdPrefix(YarnConfiguration config) {
        return "bash";
    }

    @Override
    public String cmdContainerExtra(String cmd, DtYarnConfiguration conf, Map<String, Object> containerInfo) {
        String jupyterProject = conf.get(JUPYTER_PROJECT_DIR_KEY);
        File jupyterProjectDir = new File(jupyterProject);
        if (!jupyterProjectDir.exists()) {
            if (!jupyterProjectDir.mkdirs()) {
                throw new RuntimeException("create the jupyterWorkspaceRootDir of " + jupyterProject + " failed");
            }
        }

        String[] configPort = conf.getStrings(DtYarnConfiguration.APP_CONTAINER_PORT_RANGE, DEFAULT_PORT_RANGE);
        int portStart = Integer.valueOf(DEFAULT_PORT_RANGE[0]);
        int portEnd = Integer.valueOf(DEFAULT_PORT_RANGE[1]);
        if (configPort.length == 1) {
            portStart = Integer.valueOf(configPort[0]);
        } else if (configPort.length >= DEFAULT_PORT_RANGE.length) {
            portStart = Integer.valueOf(configPort[0]);
            portEnd = Integer.valueOf(configPort[1]);
        }
        int port = NetUtils.getAvailablePortRange(portStart, portEnd);
        containerInfo.put("port", port);
        return cmd.replace(JUPYTER_NOTEBOOK_CONFIG_PORT_TMP, String.valueOf(port));
    }

    private String generateJupyterNotebookConfig(YarnConfiguration conf, String workspace, String baseUrl) {
        StringBuilder jncsb = new StringBuilder(200);
        jncsb.append("cat > ").append(JUPYTER_CONFIG_FILENAME).append(" <<EOF").append(LINEFEED);
        Iterator<Map.Entry<String, String>> confEntryIt = conf.iterator();
        while (confEntryIt.hasNext()) {
            Map.Entry<String, String> confEntry = confEntryIt.next();
            if (confEntry.getKey().startsWith(JUPYTER_NOTEBOOK_CONFIG_PREFIX)) {
                jncsb.append(confEntry.getKey()).append(JUPYTER_NOTEBOOK_CONFIG_SIGN).append(confEntry.getValue()).append(LINEFEED);
            }
        }
        jncsb.append("c.NotebookApp.notebook_dir").append(JUPYTER_NOTEBOOK_CONFIG_SIGN).append("'").append(workspace).append("'").append(LINEFEED);
        jncsb.append("c.NotebookApp.base_url").append(JUPYTER_NOTEBOOK_CONFIG_SIGN).append("'").append(baseUrl).append("'").append(LINEFEED);
        jncsb.append("c.NotebookApp.port").append(JUPYTER_NOTEBOOK_CONFIG_SIGN).append(JUPYTER_NOTEBOOK_CONFIG_PORT_TMP).append(LINEFEED);
        jncsb.append("EOF").append(LINEFEED);
        return jncsb.toString();
    }

    @Override
    public String name() {
        return "Jupyter";
    }

    @Override
    public void env(List<String> envList) {
        super.env(envList);
    }

}