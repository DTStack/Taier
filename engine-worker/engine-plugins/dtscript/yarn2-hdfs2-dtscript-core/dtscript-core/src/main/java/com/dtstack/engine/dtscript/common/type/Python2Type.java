package com.dtstack.engine.dtscript.common.type;


import com.dtstack.engine.dtscript.DtYarnConfiguration;
import org.apache.commons.lang.StringUtils;
import org.apache.hadoop.yarn.conf.YarnConfiguration;


public class Python2Type extends AbstractPythonType {

    @Override
    public String cmdPrefix(YarnConfiguration config) {
        String python = config.get(DtYarnConfiguration.PYTHON2_PATH);
        return StringUtils.isNotBlank(python) ? python : "python";
    }

    @Override
    public String name() {
        return "PYTHON2";
    }


}