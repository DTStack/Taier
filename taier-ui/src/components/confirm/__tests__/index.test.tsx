import type { CatalogueDataProps } from '@/interface';
import { act, cleanup, fireEvent, render } from '@testing-library/react';
import Confirm, { confirm } from '..';

describe('Test Confirm Component', () => {
    beforeEach(() => {
        cleanup();
        document.body.innerHTML = '';
    });

    describe('Test Inner Confirm', () => {
        it('Should match snapshot', () => {
            const { asFragment } = render(<Confirm open tab={{ name: 'test' } as CatalogueDataProps} />);
            expect(asFragment()).toMatchSnapshot();
        });

        it('Should focus on save button', () => {
            const { container } = render(<Confirm open tab={{ name: 'test' } as CatalogueDataProps} />);

            expect(document.activeElement).toEqual(container.querySelectorAll('button').item(0));
        });
    });

    it('Should call confirm with function', () => {
        const saveFn = jest.fn();
        const unSaveFn = jest.fn();
        const cancelFn = jest.fn();

        act(() => {
            confirm({
                tab: { name: 'test' } as CatalogueDataProps,
                onSave: saveFn,
                onUnSave: unSaveFn,
                onCancel: cancelFn,
            });
        });

        expect(document.body).toMatchSnapshot();

        const buttons = document.body.querySelectorAll('button');

        act(() => {
            fireEvent.click(buttons[0]);
        });
        expect(saveFn).toBeCalledTimes(1);

        act(() => {
            confirm({
                tab: { name: 'test' } as CatalogueDataProps,
                onSave: saveFn,
                onUnSave: unSaveFn,
                onCancel: cancelFn,
            });
        });
        act(() => {
            fireEvent.click(buttons[1]);
        });
        expect(unSaveFn).toBeCalledTimes(1);

        act(() => {
            confirm({
                tab: { name: 'test' } as CatalogueDataProps,
                onSave: saveFn,
                onUnSave: unSaveFn,
                onCancel: cancelFn,
            });
        });
        act(() => {
            fireEvent.click(buttons[2]);
        });
        expect(cancelFn).toBeCalledTimes(1);
    });
});
