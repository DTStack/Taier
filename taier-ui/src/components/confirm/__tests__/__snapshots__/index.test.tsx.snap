// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Confirm Component Should call confirm with function 1`] = `
<body>
  <div
    class="taier--confirm"
  >
    <p
      data-testid="antd-mock-Modal-title"
    />
    <span
      aria-label="warning"
      class="anticon anticon-warning taier__confirm__icon"
      role="img"
    >
      <svg
        aria-hidden="true"
        data-icon="warning"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="64 64 896 896"
        width="1em"
      >
        <path
          d="M464 720a48 48 0 1096 0 48 48 0 10-96 0zm16-304v184c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V416c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8zm475.7 440l-416-720c-6.2-10.7-16.9-16-27.7-16s-21.6 5.3-27.7 16l-416 720C56 877.4 71.4 904 96 904h832c24.6 0 40-26.6 27.7-48zm-783.5-27.9L512 239.9l339.8 588.2H172.2z"
        />
      </svg>
    </span>
    <div
      class="taier__confirm__title"
    >
      是否要保存对 
      test
       的修改
    </div>
    <div
      class="taier__confirm__desc"
    >
      如果不保存，将丢失修改
    </div>
    <div
      class="taier__confirm__btnGroups"
    >
      <button
        class="ant-btn ant-btn-primary ant-btn-block"
        type="button"
      >
        <span>
          保 存
        </span>
      </button>
      <button
        class="ant-btn ant-btn-ghost ant-btn-block"
        type="button"
      >
        <span>
          不保存
        </span>
      </button>
      <button
        class="ant-btn ant-btn-ghost ant-btn-block"
        type="button"
      >
        <span>
          取 消
        </span>
      </button>
    </div>
    <button
      data-testid="antd-mock-Modal-confirm"
    >
      confirm
    </button>
  </div>
</body>
`;

exports[`Test Confirm Component Test Inner Confirm Should match snapshot 1`] = `
<DocumentFragment>
  <p
    data-testid="antd-mock-Modal-title"
  />
  <span
    aria-label="warning"
    class="anticon anticon-warning taier__confirm__icon"
    role="img"
  >
    <svg
      aria-hidden="true"
      data-icon="warning"
      fill="currentColor"
      focusable="false"
      height="1em"
      viewBox="64 64 896 896"
      width="1em"
    >
      <path
        d="M464 720a48 48 0 1096 0 48 48 0 10-96 0zm16-304v184c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V416c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8zm475.7 440l-416-720c-6.2-10.7-16.9-16-27.7-16s-21.6 5.3-27.7 16l-416 720C56 877.4 71.4 904 96 904h832c24.6 0 40-26.6 27.7-48zm-783.5-27.9L512 239.9l339.8 588.2H172.2z"
      />
    </svg>
  </span>
  <div
    class="taier__confirm__title"
  >
    是否要保存对 test 的修改
  </div>
  <div
    class="taier__confirm__desc"
  >
    如果不保存，将丢失修改
  </div>
  <div
    class="taier__confirm__btnGroups"
  >
    <button
      class="ant-btn ant-btn-primary ant-btn-block"
      type="button"
    >
      <span>
        保 存
      </span>
    </button>
    <button
      class="ant-btn ant-btn-ghost ant-btn-block"
      type="button"
    >
      <span>
        不保存
      </span>
    </button>
    <button
      class="ant-btn ant-btn-ghost ant-btn-block"
      type="button"
    >
      <span>
        取 消
      </span>
    </button>
  </div>
  <button
    data-testid="antd-mock-Modal-confirm"
  >
    confirm
  </button>
</DocumentFragment>
`;
